crates/core/component/stake/src/penalty.rs

```diff
@@ -99,80 +132,17 @@ impl AllocVar<Penalty, Fq> for PenaltyVar {
         f: impl FnOnce() -> Result<T, SynthesisError>,
         mode: ark_r1cs_std::prelude::AllocationMode,
     ) -> Result<Self, SynthesisError> {
-        let ns = cs.into();
-        let cs = ns.cs();
-        let inner: Penalty = *f()?.borrow();
-        let penalty = FqVar::new_variable(cs, || Ok(Fq::from(inner.0)), mode)?;
-        // Check the Penalty is 64 bits
-        let _ = bit_constrain(penalty.clone(), 64);
-        Ok(Self { inner: penalty })
-    }
-}
-
-impl From<&PenaltyVar> for AmountVar {
-    fn from(penalty_var: &PenaltyVar) -> Self {
-        // `AmountVar`s must fit in 128 bits, but `PenaltyVar`s have already
-        // been constrained to fit in 64 bits, so we can safely
-        // construct an `AmountVar` from a `PenaltyVar`.
-        AmountVar {
-            amount: penalty_var.inner.clone(),
-        }
+        Ok(Self {
```

Comment:
Here in `AllocVar` as an additional guard we could return an error if `mode` is not `AllocationMode::Input`